### PR TITLE
systemd: Make build options configurable

### PR DESCRIPTION
**Pull request by jameysharp**
###### Motivation for this change

The binary closure for systemd is quite large in the default configuration. This patch provides a simple way to turn off build options that aren't needed by some users, removing the corresponding dependencies. It also makes it easy to turn on additional features that NixOS doesn't enable by default.

Since `doCheck` was disabled, this commit also disables building tests by default. If the caller requests building tests, then `doCheck` will be turned on as well. The tests are probably still just as broken as they were before, but if someone wants to try them, they can build with `features.tests=true`.

###### Things done

The build log from this version is identical to the build log from the previous version, with the following exceptions:

- Store paths vary, of course.
- Configure messages like "Dependency libselinux found: NO" are gone; the corresponding checks are explicitly disabled now.
- Tests aren't built since we don't run them.
- The build phase compiles files in different orders due to parallel build.
- The install and fixup phases touch files in different orders, but they touch all the same files.

The `man` output's closure size is 8 bytes bigger (a less gzip'able store path in a man page somewhere, I guess?) but the other outputs have identical closure sizes.

Also, the same set of paths appear in each output before and after this patch. Their contents differ due to varying store paths so I haven't checked if there are more subtle differences, but based on the above results I suspect they're equivalent.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/systemd/default.nix [hunk 1]
```diff
@@ -46,79 +216,39 @@ in stdenv.mkDerivation rec {
...
+    mesonFlagsArray+=(${lib.concatStringsSep " " [
+      (lib.escapeShellArgs (lib.mapAttrsToList featureFlag finalFeatures))
+      "-Ddbuspolicydir=$out/etc/dbus-1/system.d"
+      "-Ddbussessionservicedir=$out/share/dbus-1/services"
+      "-Ddbussystemservicedir=$out/share/dbus-1/system-services"
+      "-Dpamconfdir=$out/etc/pam.d"
+      "-Dpkgconfigdatadir=$dev/share/pkgconfig"
+      "-Dpkgconfiglibdir=$dev/lib/pkgconfig"
+      "-Drootlibdir=$lib/lib"
+      "-Drootprefix=$out"
```
Comment (Mic92) [inline pkgs/os-specific/linux/systemd/default.nix]: using placeholder of nix2 this could be moved to `mesonFlags` in order to avoid escaping.

Comment (jameysharp) [reply pkgs/os-specific/linux/systemd/default.nix]: Thanks for reviewing! I don't know what that means though. Can you give an example or point me at the right documentation?

Comment (ma9e) [reply pkgs/os-specific/linux/systemd/default.nix]: Derivation attributes can now reference the outputs of the derivation using the placeholder builtin function. For example, the attribute
```
configureFlags = "--prefix=${placeholder "out"} --includedir=${placeholder "dev"}";
```
will cause the configureFlags environment variable to contain the actual store paths corresponding to the out and dev outputs.

**Comment by aanderse** [comment]
@jameysharp any motivation to continue with this PR? From other PRs of yours I'm assuming this PR still has significant value to you.

**Comment by jameysharp** [comment]
Yes! I've been meaning to get back to it. I think the comments about using `placeholder` got resolved on master already so I just need to rebase and re-test now that master has a new version of systemd.

File: pkgs/os-specific/linux/systemd/default.nix [hunk 2]
```diff
@@ -1,22 +1,192 @@
...
+      "system-gid-max" = 499;
+      "ntp-servers" = "0.nixos.pool.ntp.org 1.nixos.pool.ntp.org 2.nixos.pool.ntp.org 3.nixos.pool.ntp.org";
+      # "time-epoch" = 1;
+
+      # NixOS doesn't need SysV init compatibility.
+      "sysvinit-path" = "";
+      "sysvrcnd-path" = "";
+
+      # Needed if the gnu-efi feature is enabled; ignored otherwise.
+      "efi-includedir" = "${toString gnu-efi}/include/efi";
```
Comment (doronbehar) [inline pkgs/os-specific/linux/systemd/default.nix]: Why do you need `toString`?

**Comment by doronbehar** [comment]
I like this PR as well and wish it would get merged.

**Comment by doronbehar** [comment]
Also, I'd vote for adding another build flavor for systemd in `all-packages.nix` perhaps to be named `systemdLite` and/or `systemdFull` with less / more features enabled / disabled.

File: pkgs/os-specific/linux/systemd/default.nix (hunk 2 above)
Comment (jameysharp) [reply pkgs/os-specific/linux/systemd/default.nix:164]: I think you're right that this doesn't need `toString`, but I copied those settings directly from the existing code and I didn't want to change anything like that. Those `toString` calls are still in the current version of the package, even.

**Comment by flokli** [comment]
Sorry, seems I missed this PR. Has this been solved by https://github.com/NixOS/nixpkgs/pull/98998 (and the subsequent PRs)?

**Comment by flokli** [comment]
Let's close this. It has probably been addressed in other PRs in the meantime, and there was no reaction on https://github.com/NixOS/nixpkgs/pull/53222#issuecomment-751255463.

**Pull request closed without merging**